src/umockdev.vala

```diff
@@ -687,6 +714,50 @@ public class Testbed: GLib.Object {
         remove_dir(real_path, true);
     }
 
+    private string? find_a_child(string syspath)
+    {
+      try {
+        string real_path = Path.build_filename(this.root_dir, syspath);
+        Dir dir = Dir.open(real_path);
+        string? name = null;
+        while ((name = dir.read_name()) != null) {
+          string path = Path.build_filename(real_path, name);
+          // Skip over symlinks
+          if (FileUtils.test(path, FileTest.IS_SYMLINK)) {
+            continue;
+          }
+          if (FileUtils.test(path, FileTest.IS_REGULAR)) {
+            if (name == "uevent") {
+              return syspath;
+            }
+          }
+        }
+
+        // If we're here, then this path is not a device, but it might
+        // have a child device hidden deeper.
```

Comment:
This needs to happen *first*, i.e. the search needs to be depth-first. You need to remove the children first before the parents. This doesn't matter much in terms of removing the files (as that'll happen at the very end through `remove_dir()`, but it'll produce the wrong (inverted) order of `remove` events.

Reply:
It actually didn't invert the order, but you are correct that it is confusing. I will rework it per your recommendation above.